File: platform/daemon/src/auth/policy.ts
/**
 * Permission matrix and scope enforcement.
 */

import type { AuthMode, Permission, PolicyDecision, TokenClaims, TokenRole, TokenScope } from "./types";
import { logger } from "../logger";

// Track which subs have been warned about empty scope to avoid log flooding.
// Unbounded for the process lifetime — acceptable for typical deployments
// where the number of distinct token subjects is small. If ephemeral per-session
// subs are used at high volume, consider replacing with a bounded LRU.
const warnedEmptyScope = new Set<string>();

const PERMISSION_MATRIX: Readonly<Record<TokenRole, readonly Permission[]>> = {
	admin: [
		"remember",
		"recall",
		"modify",
		"forget",
		"recover",
		"admin",
		"documents",
		"connectors",
		"diagnostics",
		"analytics",
	],
	operator: [
		"remember",
		"recall",
		"modify",
		"forget",
		"recover",
		"documents",
		"connectors",
		"diagnostics",
		"analytics",
	],
	agent: ["remember", "recall", "modify", "forget", "recover", "documents"],
	readonly: ["recall"],
};

const permissionSets = new Map<TokenRole, ReadonlySet<Permission>>(
	Object.entries(PERMISSION_MATRIX).map(([role, perms]) => [role as TokenRole, new Set(perms)]),
);

export function checkPermission(
	claims: TokenClaims | null,
	permission: Permission,
	authMode: AuthMode,
): PolicyDecision {
	if (authMode === "local") {
		return { allowed: true };
	}

	if (!claims) {
		return { allowed: false, reason: "authentication required" };
	}

	const allowed = permissionSets.get(claims.role);
	if (!allowed || !allowed.has(permission)) {
		return {
			allowed: false,
			reason: `role '${claims.role}' lacks '${permission}' permission`,
		};
	}

	if (claims.permissions && !claims.permissions.includes(permission)) {
		return {
			allowed: false,
			reason: `credential lacks '${permission}' permission`,
		};
	}

	return { allowed: true };
}

export function checkScope(claims: TokenClaims | null, target: TokenScope, authMode: AuthMode): PolicyDecision {
	if (authMode === "local") {
		return { allowed: true };
	}

	if (!claims) {
		return { allowed: false, reason: "authentication required" };
	}

	// Admin role bypasses scope checks
	if (claims.role === "admin") {
		return { allowed: true };
	}

	// DEPRECATION: Non-admin tokens with empty scope currently get full access
	// but will be denied in a future release. Log once per sub to avoid
	// flooding structured logs on busy deployments.
	const scope = claims.scope;
	if (!scope.project && !scope.agent && !scope.user) {
		if (!warnedEmptyScope.has(claims.sub)) {
			warnedEmptyScope.add(claims.sub);
			logger.warn(
				"daemon",
				"DEPRECATION: non-admin token has empty scope — will be denied in a future release. Issue tokens with explicit scope fields.",
				{
					sub: claims.sub,
					role: claims.role,
				},
			);
		}
		return { allowed: true };
	}

	if (scope.project && target.project && scope.project !== target.project) {
		return {
			allowed: false,
			reason: `scope restricted to project '${scope.project}'`,
		};
	}

	if (scope.agent && target.agent && scope.agent !== target.agent) {
		return {
			allowed: false,
			reason: `scope restricted to agent '${scope.agent}'`,
		};
	}

	if (scope.user && target.user && scope.user !== target.user) {
		return {
			allowed: false,
			reason: `scope restricted to user '${scope.user}'`,
		};
	}

	return { allowed: true };
}

export { PERMISSION_MATRIX };

File: platform/daemon-rs/crates/signet-daemon/src/auth/policy.rs
//! Permission matrix and scope enforcement.

use super::types::{AuthMode, Permission, PolicyDecision, TokenClaims, TokenRole, TokenScope};

// ---------------------------------------------------------------------------
// Permission matrix
// ---------------------------------------------------------------------------

fn role_has_permission(role: TokenRole, perm: Permission) -> bool {
    match role {
        TokenRole::Admin => true,
        TokenRole::Operator => !matches!(perm, Permission::Admin),
        TokenRole::Agent => matches!(
            perm,
            Permission::Remember
                | Permission::Recall
                | Permission::Modify
                | Permission::Forget
                | Permission::Recover
                | Permission::Documents
        ),
        TokenRole::Readonly => matches!(perm, Permission::Recall),
    }
}

fn role_name(role: TokenRole) -> &'static str {
    match role {
        TokenRole::Admin => "admin",
        TokenRole::Operator => "operator",
        TokenRole::Agent => "agent",
        TokenRole::Readonly => "readonly",
    }
}

fn permission_name(permission: Permission) -> &'static str {
    match permission {
        Permission::Remember => "remember",
        Permission::Recall => "recall",
        Permission::Modify => "modify",
        Permission::Forget => "forget",
        Permission::Recover => "recover",
        Permission::Admin => "admin",
        Permission::Documents => "documents",
        Permission::Connectors => "connectors",
        Permission::Diagnostics => "diagnostics",
        Permission::Analytics => "analytics",
    }
}

// ---------------------------------------------------------------------------
// Permission check
// ---------------------------------------------------------------------------

pub fn check_permission(
    claims: Option<&TokenClaims>,
    perm: Permission,
    mode: AuthMode,
) -> PolicyDecision {
    if mode == AuthMode::Local {
        return PolicyDecision {
            allowed: true,
            reason: None,
        };
    }

    let Some(claims) = claims else {
        return PolicyDecision {
            allowed: false,
            reason: Some("authentication required".into()),
        };
    };

    if !role_has_permission(claims.role, perm) {
        return PolicyDecision {
            allowed: false,
            reason: Some(format!(
                "role '{}' lacks '{}' permission",
                role_name(claims.role),
                permission_name(perm)
            )),
        };
    }

    if let Some(permissions) = claims.permissions.as_ref()
        && !permissions.contains(&perm)
    {
        return PolicyDecision {
            allowed: false,
            reason: Some(format!(
                "credential lacks '{}' permission",
                permission_name(perm)
            )),
        };
    }

    PolicyDecision {
        allowed: true,
        reason: None,
    }
}

// ---------------------------------------------------------------------------
// Scope check
// ---------------------------------------------------------------------------

pub fn check_scope(
    claims: Option<&TokenClaims>,
    target: &TokenScope,
    mode: AuthMode,
) -> PolicyDecision {
    if mode == AuthMode::Local {
        return PolicyDecision {
            allowed: true,
            reason: None,
        };
    }

    let Some(claims) = claims else {
        return PolicyDecision {
            allowed: false,
            reason: Some("authentication required".into()),
        };
    };

    // Admin bypasses scope
    if claims.role == TokenRole::Admin {
        return PolicyDecision {
            allowed: true,
            reason: None,
        };
    }

    // Unscoped tokens have full access
    if claims.scope.is_empty() {
        return PolicyDecision {
            allowed: true,
            reason: None,
        };
    }

    if let (Some(sp), Some(tp)) = (&claims.scope.project, &target.project)
        && sp != tp
    {
        return PolicyDecision {
            allowed: false,
            reason: Some(format!("scope restricted to project '{sp}'")),
        };
    }

    if let (Some(sa), Some(ta)) = (&claims.scope.agent, &target.agent)
        && sa != ta
    {
        return PolicyDecision {
            allowed: false,
            reason: Some(format!("scope restricted to agent '{sa}'")),
        };
    }

    if let (Some(su), Some(tu)) = (&claims.scope.user, &target.user)
        && su != tu
    {
        return PolicyDecision {
            allowed: false,
            reason: Some(format!("scope restricted to user '{su}'")),
        };
    }

    PolicyDecision {
        allowed: true,
        reason: None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn claims(role: TokenRole) -> TokenClaims {
        TokenClaims {
            sub: "test".into(),
            scope: TokenScope::default(),
            role,
            iat: 0,
            exp: i64::MAX,
            permissions: None,
        }
    }

    #[test]
    fn local_mode_allows_all() {
        let d = check_permission(None, Permission::Admin, AuthMode::Local);
        assert!(d.allowed);
    }

    #[test]
    fn admin_has_all_permissions() {
        let c = claims(TokenRole::Admin);
        for perm in [
            Permission::Remember,
            Permission::Recall,
            Permission::Admin,
            Permission::Diagnostics,
        ] {
            let d = check_permission(Some(&c), perm, AuthMode::Team);
            assert!(d.allowed, "{perm:?} should be allowed for admin");
        }
    }

    #[test]
    fn readonly_only_recall() {
        let c = claims(TokenRole::Readonly);
        assert!(check_permission(Some(&c), Permission::Recall, AuthMode::Team).allowed);
        assert!(!check_permission(Some(&c), Permission::Remember, AuthMode::Team).allowed);
        assert!(!check_permission(Some(&c), Permission::Admin, AuthMode::Team).allowed);
    }

    #[test]
    fn agent_no_admin() {
        let c = claims(TokenRole::Agent);
        assert!(check_permission(Some(&c), Permission::Remember, AuthMode::Team).allowed);
        assert!(!check_permission(Some(&c), Permission::Admin, AuthMode::Team).allowed);
        assert!(!check_permission(Some(&c), Permission::Diagnostics, AuthMode::Team).allowed);
    }

    #[test]
    fn scope_enforcement() {
        let mut c = claims(TokenRole::Agent);
        c.scope.project = Some("proj-a".into());

        let target = TokenScope {
            project: Some("proj-b".into()),
            ..Default::default()
        };
        assert!(!check_scope(Some(&c), &target, AuthMode::Team).allowed);

        let target = TokenScope {
            project: Some("proj-a".into()),
            ..Default::default()
        };
        assert!(check_scope(Some(&c), &target, AuthMode::Team).allowed);
    }

    #[test]
    fn admin_bypasses_scope() {
        let mut c = claims(TokenRole::Admin);
        c.scope.project = Some("proj-a".into());

        let target = TokenScope {
            project: Some("proj-b".into()),
            ..Default::default()
        };
        assert!(check_scope(Some(&c), &target, AuthMode::Team).allowed);
    }
}
